### PR TITLE
Add key constraints for the JoinOrder benchmark

### PR DESCRIPTION
**Pull request by julianmenzler**
Currently, we do not really profit from key constraints in the JOB benchmark. But, this may change in the future.

<details>
<summary>Configuration Overview - click to expand</summary>

```diff
 +Configuration Overview---------------------------------------------+------------------------------------------------+
 | Parameter        | job_master_80e13c.json                         | job_add_constraints_b5153b9.json               |
 +------------------+------------------------------------------------+------------------------------------------------+
 |  GIT-HASH        | 80e13c23d433b8b0c4b562084def804b2c94d415-dirty | b5153b9c0d7b2bb01123830110333332093964e2-dirty |
 |  benchmark_mode  | Ordered                                        | Ordered                                        |
 |  build_type      | release                                        | release                                        |
 |  chunk_size      | 65535                                          | 65535                                          |
 |  clients         | 1                                              | 1                                              |
 |  compiler        | gcc 9.2                                        | gcc 9.2                                        |
 |  cores           | 0                                              | 0                                              |
 |  date            | 2021-06-09 13:21:04                            | 2021-06-10 19:31:23                            |
 |  encoding        | {'default': {'encoding': 'Dictionary'}}        | {'default': {'encoding': 'Dictionary'}}        |
 |  indexes         | False                                          | False                                          |
 |  max_duration    | 60000000000                                    | 60000000000                                    |
 |  max_runs        | -1                                             | -1                                             |
 |  time_unit       | ns                                             | ns                                             |
 |  using_scheduler | False                                          | False                                          |
 |  verify          | False                                          | False                                          |
 |  warmup_duration | 0                                              | 0                                              |
 +------------------+------------------------------------------------+------------------------------------------------+
```
</details>

```diff
 +---------++----------+----------+--------++----------+----------+--------+---------+
 | Item    || Latency (ms/iter)   | Change || Throughput (iter/s) | Change | p-value |
 |         ||      old |      new |        ||      old |      new |        |         |
 +---------++----------+----------+--------++----------+----------+--------+---------+
 | 10a     ||    737.7 |    734.4 |   -0%  ||     1.36 |     1.36 |   +0%  |  0.4487 |
 | 10b     ||    715.6 |    714.9 |   -0%  ||     1.40 |     1.40 |   +0%  |  0.7305 |
+| 10c     ||   1651.4 |   1379.8 |  -16%  ||     0.61 |     0.72 |  +20%  |  0.0000 |
 | 11a     ||    269.9 |    265.8 |   -2%  ||     3.70 |     3.76 |   +2%  |  0.3543 |
 | 11b     ||    255.9 |    253.4 |   -1%  ||     3.91 |     3.95 |   +1%  |  0.5533 |
+| 11c     ||    999.1 |    897.5 |  -10%  ||     1.00 |     1.11 |  +11%  |  0.0000 |
+| 11d     ||  10667.9 |   9523.2 |  -11%  ||     0.09 |     0.11 |  +12%  |       ˅ |
+| 12a     ||    587.2 |    547.9 |   -7%  ||     1.70 |     1.83 |   +7%  |  0.0001 |
 | 12b     ||    354.3 |    354.5 |   +0%  ||     2.82 |     2.82 |   -0%  |  0.9775 |
 | 12c     ||   6688.3 |   6737.5 |   +1%  ||     0.15 |     0.15 |   -1%  |       ˅ |
-| 13a     ||    332.8 |    454.5 |  +37%  ||     3.00 |     2.20 |  -27%  |  0.0000 |
+| 13b     ||    424.0 |    404.8 |   -5%  ||     2.36 |     2.47 |   +5%  |  0.0000 |
+| 13c     ||    320.6 |    304.8 |   -5%  ||     3.12 |     3.28 |   +5%  |  0.0000 |
-| 13d     ||    332.5 |    453.7 |  +36%  ||     3.01 |     2.20 |  -27%  |  0.0000 |
 | 14a     ||   8012.3 |   7954.7 |   -1%  ||     0.12 |     0.13 |   +1%  |       ˅ |
 | 14b     ||   8360.9 |   8308.0 |   -1%  ||     0.12 |     0.12 |   +1%  |       ˅ |
 | 14c     ||   8018.9 |   7945.9 |   -1%  ||     0.12 |     0.13 |   +1%  |       ˅ |
 | 15a     ||    406.5 |    396.8 |   -2%  ||     2.46 |     2.52 |   +2%  |  0.0060 |
 | 15b     ||    394.5 |    386.1 |   -2%  ||     2.53 |     2.59 |   +2%  |  0.0004 |
 | 15c     ||    304.0 |    300.0 |   -1%  ||     3.29 |     3.33 |   +1%  |  0.1705 |
+| 15d     ||    925.6 |    749.9 |  -19%  ||     1.08 |     1.33 |  +23%  |  0.0000 |
+| 16a     ||  10151.2 |   9497.8 |   -6%  ||     0.10 |     0.11 |   +7%  |       ˅ |
+| 16b     ||  13112.9 |  12488.7 |   -5%  ||     0.08 |     0.08 |   +5%  |       ˅ |
+| 16c     ||  10406.0 |   9821.1 |   -6%  ||     0.10 |     0.10 |   +6%  |       ˅ |
+| 16d     ||  10326.5 |   9775.0 |   -5%  ||     0.10 |     0.10 |   +6%  |       ˅ |
 | 17a     ||   1879.9 |   1826.9 |   -3%  ||     0.53 |     0.55 |   +3%  |  0.0094 |
 | 17b     ||   1537.3 |   1501.2 |   -2%  ||     0.65 |     0.67 |   +2%  |  0.0195 |
 | 17c     ||   1460.7 |   1445.8 |   -1%  ||     0.68 |     0.69 |   +1%  |  0.3090 |
 | 17d     ||   1662.6 |   1639.5 |   -1%  ||     0.60 |     0.61 |   +1%  |  0.1574 |
+| 17e     ||   5371.3 |   4990.6 |   -7%  ||     0.19 |     0.20 |   +8%  |  0.0000 |
+| 17f     ||   2991.8 |   2850.1 |   -5%  ||     0.33 |     0.35 |   +5%  |  0.0001 |
-| 18a     ||   1180.6 |   1689.4 |  +43%  ||     0.85 |     0.59 |  -30%  |  0.0000 |
-| 18b     ||    529.0 |    566.9 |   +7%  ||     1.89 |     1.76 |   -7%  |  0.0000 |
 | 18c     ||   7017.4 |   6973.5 |   -1%  ||     0.14 |     0.14 |   +1%  |       ˅ |
 | 19a     ||  10565.6 |  10438.7 |   -1%  ||     0.09 |     0.10 |   +1%  |       ˅ |
 | 19b     ||   4963.7 |   4892.5 |   -1%  ||     0.20 |     0.20 |   +1%  |  0.0144 |
 | 19c     ||  10224.0 |  10094.9 |   -1%  ||     0.10 |     0.10 |   +1%  |       ˅ |
+| 19d     ||  10832.3 |   9426.4 |  -13%  ||     0.09 |     0.11 |  +15%  |       ˅ |
 | 1a      ||    114.9 |    113.4 |   -1%  ||     8.70 |     8.82 |   +1%  |  0.0000 |
 | 1b      ||     47.4 |     47.5 |   +0%  ||    21.07 |    21.05 |   -0%  |  0.0356 |
 | 1c      ||     48.2 |     48.1 |   -0%  ||    20.76 |    20.80 |   +0%  |  0.0035 |
 | 1d      ||     47.4 |     47.5 |   +0%  ||    21.09 |    21.07 |   -0%  |  0.0978 |
 | 20a     ||   2209.3 |   2167.3 |   -2%  ||     0.45 |     0.46 |   +2%  |  0.0000 |
 | 20b     ||   1894.0 |   1862.5 |   -2%  ||     0.53 |     0.54 |   +2%  |  0.0000 |
 | 20c     ||   1231.6 |   1200.5 |   -3%  ||     0.81 |     0.83 |   +3%  |  0.0000 |
 | 21a     ||   7406.4 |   7366.5 |   -1%  ||     0.14 |     0.14 |   +1%  |       ˅ |
-| 21b     ||    411.2 |    436.3 |   +6%  ||     2.43 |     2.29 |   -6%  |  0.0000 |
 | 21c     ||   7704.5 |   7676.3 |   -0%  ||     0.13 |     0.13 |   +0%  |       ˅ |
 | 22a     ||   6378.4 |   6274.1 |   -2%  ||     0.16 |     0.16 |   +2%  |  0.2469 |
 | 22b     ||   6369.2 |   6277.0 |   -1%  ||     0.16 |     0.16 |   +1%  |  0.3064 |
 | 22c     ||   8035.7 |   7956.0 |   -1%  ||     0.12 |     0.13 |   +1%  |       ˅ |
 | 22d     ||   8071.1 |   7981.8 |   -1%  ||     0.12 |     0.13 |   +1%  |       ˅ |
 | 23a     ||    303.9 |    297.1 |   -2%  ||     3.29 |     3.37 |   +2%  |  0.0016 |
 | 23b     ||    221.6 |    218.0 |   -2%  ||     4.51 |     4.59 |   +2%  |  0.0129 |
 | 23c     ||    323.0 |    317.0 |   -2%  ||     3.10 |     3.15 |   +2%  |  0.0071 |
 | 24a     ||   5238.1 |   5056.1 |   -3%  ||     0.19 |     0.20 |   +4%  |  0.0211 |
 | 24b     ||   5086.2 |   4961.8 |   -2%  ||     0.20 |     0.20 |   +3%  |  0.3382 |
 | 25a     ||    533.5 |    530.9 |   -0%  ||     1.87 |     1.88 |   +0%  |  0.4391 |
 | 25b     ||    558.1 |    557.8 |   -0%  ||     1.79 |     1.79 |   +0%  |  0.9080 |
 | 25c     ||   7052.5 |   6979.4 |   -1%  ||     0.14 |     0.14 |   +1%  |       ˅ |
 | 26a     ||   1037.1 |   1031.2 |   -1%  ||     0.96 |     0.97 |   +1%  |  0.4843 |
 | 26b     ||   1010.8 |   1007.1 |   -0%  ||     0.99 |     0.99 |   +0%  |  0.6382 |
 | 26c     ||   1039.0 |   1031.7 |   -1%  ||     0.96 |     0.97 |   +1%  |  0.3893 |
 | 27a     ||   6295.0 |   6332.0 |   +1%  ||     0.16 |     0.16 |   -1%  |  0.6593 |
 | 27b     ||   6250.3 |   6202.1 |   -1%  ||     0.16 |     0.16 |   +1%  |  0.4474 |
 | 27c     ||    369.7 |    381.7 |   +3%  ||     2.71 |     2.62 |   -3%  |  0.0104 |
 | 28a     ||   8173.0 |   8173.8 |   +0%  ||     0.12 |     0.12 |   -0%  |       ˅ |
 | 28b     ||   6255.5 |   6340.8 |   +1%  ||     0.16 |     0.16 |   -1%  |  0.8001 |
 | 28c     ||   8171.7 |   8156.9 |   -0%  ||     0.12 |     0.12 |   +0%  |       ˅ |
 | 29a     ||   5156.1 |   5075.2 |   -2%  ||     0.19 |     0.20 |   +2%  |  0.8879 |
 | 29b     ||    347.4 |    345.8 |   -0%  ||     2.88 |     2.89 |   +0%  |  0.9642 |
 | 29c     ||   5387.2 |   5239.9 |   -3%  ||     0.19 |     0.19 |   +3%  |  0.7583 |
 | 2a      ||    181.5 |    174.1 |   -4%  ||     5.51 |     5.74 |   +4%  |  0.0000 |
 | 2b      ||    150.1 |    143.8 |   -4%  ||     6.66 |     6.95 |   +4%  |  0.0000 |
-| 2c      ||    102.7 |    117.7 |  +15%  ||     9.74 |     8.50 |  -13%  |  0.0000 |
+| 2d      ||    224.1 |    211.3 |   -6%  ||     4.46 |     4.73 |   +6%  |  0.0000 |
 | 30a     ||    563.6 |    552.9 |   -2%  ||     1.77 |     1.81 |   +2%  |  0.5235 |
 | 30b     ||    892.2 |    881.1 |   -1%  ||     1.12 |     1.13 |   +1%  |  0.6834 |
 | 30c     ||   7166.7 |   7076.4 |   -1%  ||     0.14 |     0.14 |   +1%  |       ˅ |
 | 31a     ||    629.5 |    625.9 |   -1%  ||     1.59 |     1.60 |   +1%  |  0.7273 |
 | 31b     ||    951.5 |    960.5 |   +1%  ||     1.05 |     1.04 |   -1%  |  0.6065 |
 | 31c     ||    636.9 |    636.6 |   -0%  ||     1.57 |     1.57 |   +0%  |  0.9718 |
 | 32a     ||     87.3 |     87.1 |   -0%  ||    11.45 |    11.48 |   +0%  |  0.0465 |
 | 32b     ||    497.9 |    479.3 |   -4%  ||     2.01 |     2.09 |   +4%  |  0.0000 |
 | 33a     ||    141.5 |    140.1 |   -1%  ||     7.06 |     7.14 |   +1%  |  0.3551 |
 | 33b     ||    140.1 |    138.4 |   -1%  ||     7.14 |     7.22 |   +1%  |  0.2803 |
 | 33c     ||    163.1 |    158.3 |   -3%  ||     6.13 |     6.32 |   +3%  |  0.0082 |
 | 3a      ||   7578.7 |   7528.8 |   -1%  ||     0.13 |     0.13 |   +1%  |       ˅ |
 | 3b      ||     69.3 |     69.1 |   -0%  ||    14.43 |    14.48 |   +0%  |  0.0000 |
 | 3c      ||   9619.4 |   9555.9 |   -1%  ||     0.10 |     0.10 |   +1%  |       ˅ |
+| 4a      ||    251.2 |    233.9 |   -7%  ||     3.98 |     4.27 |   +7%  |  0.0000 |
 | 4b      ||     53.8 |     52.9 |   -2%  ||    18.58 |    18.90 |   +2%  |  0.0000 |
+| 4c      ||    287.6 |    266.1 |   -7%  ||     3.48 |     3.76 |   +8%  |  0.0000 |
 | 5a      ||   7285.7 |   7249.1 |   -1%  ||     0.14 |     0.14 |   +1%  |       ˅ |
 | 5b      ||    484.9 |    480.8 |   -1%  ||     2.06 |     2.08 |   +1%  |  0.0004 |
 | 5c      ||   8443.3 |   8387.0 |   -1%  ||     0.12 |     0.12 |   +1%  |       ˅ |
 | 6a      ||    522.8 |    522.8 |   -0%  ||     1.91 |     1.91 |   +0%  |  0.9919 |
 | 6b      ||   1935.8 |   1916.1 |   -1%  ||     0.52 |     0.52 |   +1%  |  0.0040 |
 | 6c      ||    524.3 |    521.9 |   -0%  ||     1.91 |     1.92 |   +0%  |  0.0000 |
 | 6d      ||   1916.3 |   1923.9 |   +0%  ||     0.52 |     0.52 |   -0%  |  0.1736 |
 | 6e      ||    525.4 |    532.2 |   +1%  ||     1.90 |     1.88 |   -1%  |  0.0000 |
 | 6f      ||   2906.1 |   2888.3 |   -1%  ||     0.34 |     0.35 |   +1%  |  0.0209 |
 | 7a      ||    418.4 |    418.7 |   +0%  ||     2.39 |     2.39 |   -0%  |  0.9778 |
 | 7b      ||    261.1 |    268.9 |   +3%  ||     3.83 |     3.72 |   -3%  |  0.1292 |
 | 7c      ||  23723.8 |  22740.7 |   -4%  ||     0.04 |     0.04 |   +4%  |       ˅ |
 | 8a      ||    197.4 |    194.0 |   -2%  ||     5.07 |     5.15 |   +2%  |  0.0000 |
 | 8b      ||    277.4 |    270.5 |   -2%  ||     3.61 |     3.70 |   +3%  |  0.0000 |
+| 8c      ||   7559.2 |   6752.4 |  -11%  ||     0.13 |     0.15 |  +12%  |       ˅ |
+| 8d      ||   2276.7 |   1926.4 |  -15%  ||     0.44 |     0.52 |  +18%  |  0.0000 |
 | 9a      ||   6281.7 |   6203.4 |   -1%  ||     0.16 |     0.16 |   +1%  |  0.2174 |
+| 9b      ||    571.1 |    539.6 |   -6%  ||     1.75 |     1.85 |   +6%  |  0.0000 |
 | 9c      ||   6478.7 |   6313.2 |   -3%  ||     0.15 |     0.16 |   +3%  |  0.0088 |
 | 9d      ||   7664.9 |   7467.4 |   -3%  ||     0.13 |     0.13 |   +3%  |       ˅ |
 +---------++----------+----------+--------++----------+----------+--------+---------+
 | Sum     || 384395.5 | 374286.0 |   -3%  ||          |          |        |         |
 | Geomean ||          |          |        ||          |          |   +1%  |         |
 +---------++----------+----------+--------++----------+----------+--------+---------+
 |         || ˅ Insufficient number of runs for p-value calculation                  |
 +---------++----------+----------+--------++----------+----------+--------+---------+
```

### CODE REVIEW

**Review by Bensk1 — CHANGES_REQUESTED** [review]

**Comment by Bensk1** [comment]
Some queries lose performance. Is this due to the fact that the constraints enable JoinToSemiJoinRewrites (ColumnPruningRule) and these SemiJoins prevent some SemiJoinReductions?

File: src/benchmark/join_order_benchmark.cpp [hunk 1]
```diff
@@ -21,6 +21,88 @@
 using namespace opossum;               // NOLINT
 using namespace std::string_literals;  // NOLINT
 
+/**
+ * Each of the 20 JOB tables has one surrogate key. This function registers key constraints for all of them.
```
[the remark this answers is not in the document]
Comment (julianmenzler) [reply src/benchmark/join_order_benchmark.cpp:25]: You are right, `name` is missing. Thanks for catching this!

**Comment by Bensk1** [comment]
Probably I was the one missing the name table in the first place :)

Am 10.06.2021 um 18:59 schrieb Julian Menzler ***@***.***>:

﻿

@julianmenzler commented on this pull request.

________________________________

In src/benchmark/join_order_benchmark.cpp<https://github.com/hyrise/hyrise/pull/2372#discussion_r649367014>:

> @@ -21,6 +21,88 @@
 using namespace opossum;               // NOLINT
 using namespace std::string_literals;  // NOLINT

+/**
+ * Each of the 20 JOB tables has one surrogate key. This function registers key constraints for all of them.


You are right, name is missing. Thanks for catching this!

**Comment by julianmenzler** [comment]
> Some queries lose performance. Is this due to the fact that the constraints enable JoinToSemiJoinRewrites (ColumnPruningRule) and these SemiJoins prevent some SemiJoinReductions?

Not sure if there is an easy explanation that fits all queries. But in 13a and 18a, some semi join placements get worse with constraints:

* In 13a, `Semi Join where kind_id = id)` gets placed above `Inner Join where id = movie_id`, making the latter more expensive (148ms vs. 4ms)

* In 18a, `Semi Join where info_type_id = id` gets placed above `Inner Join where movie_id = movie_id`, making the latter more expensive. (341ms vs. 204ms)

**Comment by julianmenzler** [comment]
I updated the benchmark results. No real changes.

File: src/benchmark/join_order_benchmark.cpp [hunk 2]
```diff
@@ -21,6 +21,91 @@
...
+  const auto& person_info_table = table_info_by_name.at("person_info").table;
+  person_info_table->add_soft_key_constraint(
+      {{person_info_table->column_id_by_name("id")}, KeyConstraintType::PRIMARY_KEY});
+
+  const auto& role_type_table = table_info_by_name.at("role_type").table;
+  role_type_table->add_soft_key_constraint(
+      {{role_type_table->column_id_by_name("id")}, KeyConstraintType::PRIMARY_KEY});
+
+  const auto& title_table = table_info_by_name.at("title").table;
+  title_table->add_soft_key_constraint({{title_table->column_id_by_name("id")}, KeyConstraintType::PRIMARY_KEY});
```
Comment (Bouncner) [inline src/benchmark/join_order_benchmark.cpp:106]: If I look at these lines. Is there a reason not to loop over table_info_by_name and set the primary key on `id`?
Would have solved the `name` issue as well.

**Comment by Bensk1** [comment]
Currently, that would be fine. Originally, I also added foreign key constraints for some tables. As long as we don’t have them, there should be no difference.

Am 11.06.2021 um 10:04 schrieb Bouncner ***@***.***>:

﻿

@Bouncner commented on this pull request.

________________________________

In src/benchmark/join_order_benchmark.cpp<https://github.com/hyrise/hyrise/pull/2372#discussion_r649776411>:

> +  movie_link_table->add_soft_key_constraint(
+      {{movie_link_table->column_id_by_name("id")}, KeyConstraintType::PRIMARY_KEY});
+
+  const auto& name_table = table_info_by_name.at("name").table;
+  name_table->add_soft_key_constraint({{name_table->column_id_by_name("id")}, KeyConstraintType::PRIMARY_KEY});
+
+  const auto& person_info_table = table_info_by_name.at("person_info").table;
+  person_info_table->add_soft_key_constraint(
+      {{person_info_table->column_id_by_name("id")}, KeyConstraintType::PRIMARY_KEY});
+
+  const auto& role_type_table = table_info_by_name.at("role_type").table;
+  role_type_table->add_soft_key_constraint(
+      {{role_type_table->column_id_by_name("id")}, KeyConstraintType::PRIMARY_KEY});
+
+  const auto& title_table = table_info_by_name.at("title").table;
+  title_table->add_soft_key_constraint({{title_table->column_id_by_name("id")}, KeyConstraintType::PRIMARY_KEY});


If I look at these lines. Is there a reason not to loop over table_info_by_name and set the primary key on id?
Would have solved the name issue as well.

**Comment by julianmenzler** [comment]
> If I look at these lines. Is there a reason not to loop over table_info_by_name and set the primary key on id?
Would have solved the name issue as well.

To me, that sounds like a more hacky version, but it should work.

However, to stay consistent with `TPCHTableGenerator::_add_constraints`, `TPCCTableGenerator::_add_constraints` and `TPCDSTableGenerator::_add_constraints`, I would leave it as is. 
Also: With the current implementation, we can easily add other table constraints, such as FK constraints, without rewriting any code.

File: src/benchmark/join_order_benchmark.cpp (hunk 2 above)
Comment (Bouncner) [reply src/benchmark/join_order_benchmark.cpp:106]: Yeah, the FK issue you both mentioned makes sense to me.

**Review by Bensk1 — APPROVED** [review]

**Pull request merged**